docs/guide-which-query.md

```diff
@@ -14,6 +14,11 @@ possible. With this in mind, we recommend this order of priority:
    1. `getByLabelText`: Only really good for form fields, but this is the number
       one method a user finds those elements, so it should be your top
       preference.
+   1. `getByRole`: With the `name` option you can basically query the full
+      accessibility tree. It can query all the elements you can query with the
+      other accessility query. But consider that you still might want to assert
+      if an element has an actual `<label>` element or a `placeholder`
+      attribute.
```

Comment:
I think many people use `*ByText` to select `button`s, `div`s, and `span`s. Maybe we should briefly mention what the role for each of those would be and that will cover most use-cases.

Reply:
I'm not encountering much use cases to select `div` or `span` these days. These wouldn't be queryable with `getByRole` and don't have an accessible name to begin with. 

After considering the feedback in the implementation PR I would actually agree that `byText('some text', { role: 'listitem' })` would be a useful query. Simply because some roles can't have an accessible name and the text (i.e. `.textContent`) is the appropriate fallback for these.

I think I'll revert the change in the recommendation order until this is more mature (adding `byText(text, { role })` etc).

Overall I'd say we should definitely add some guard rails here when people try e.g. `getByRole('listitem', { name: 'first' })` by warning that certain roles don't have an accessible name and suggest using `byText('first', { role: 'listitem' })`. This is even a candidate for a lint rule with auto-fixing

Reply:
I changed the wording and order again. The rule of thumb is now:
- input? -> byLabelText (simple query that doesn't require much a11y knowledge. Personally I will still use `byRole('textbox', { name: '' })`. The problem is that you need to know all the various input roles but then maybe we should encourage knowing these?)
- interactive? -> byRole
- not-interactive? -> byText

Reply:
> The problem is that you need to know all the various input roles but then maybe we should encourage knowing these?

There a lot of aria roles and not necessarily one perfect role for a given use case. For example you might have a list that gets refactored to the new `feed` role later. I think the generic queries like `label` are good for use cases where you don't want to overspecify the exact roles.

Reply:
> gets refactored to the new feed role later

That isn't just a refactor. You're actually changing semantics. We shouldn't be too overzealous with our testing approach. A test should actually fail some times.

~I'm also not sure how a `feed` relates to this issue~ I just realized why this causes so much confusion. ByLabelText is actually quite problematic with this API. I was under the impression it only considers actual `label` elements. The current implementation of ByLabelText is incomplete. We should discourage ByLabelText for anything other thant `input`.

Reply:
> The current implementation of ByLabelText is incomplete

What do you mean by incomplete? It also understands `aria-label` and `aria-labelledby`, which would require a significant amount of work to implement in userspace if you are wanting to traverse labelled elements.